### ISSUE TITLE
Performance issue for Distributed TF

### ISSUE THREAD

**Issue by AlvinChen13** [tensorflow/tensorflow#4164]
Hi All,

We just enable distributed TF to train our model,  but we get very slow performance comparing to train on single machine. We just setup two machines one ps, one worker, and each machine is Xeon CPU with nvidia K40 GPU:
GPU 0: Quadro K420 (UUID: GPU-954fc4b7-6198-f5f9-1efb-69e73d001d5c)
GPU 1: Tesla K40c (UUID: GPU-ac243789-45bd-448f-87d9-db21c1c88c87)
GPU 2: Tesla K40c (UUID: GPU-d7b7783e-9041-e4e2-91f0-dc4188fd84e9)

If we train on one machine we can meet 200+examples/sec, but on two machines, it is just 20+ examples/sec. 10 times slower....

Our two machines are connected in 1GBits network, and we find the network bandwidth is just 40%.

Some topic discussed this issue, but I haven't got any solutions... 

One machine:
2016-09-02 09:04:35.330931: step 240, lr = 0.0001, loss = 1.59115 (107.4 examples/sec; 0.14900 sec/batch)
2016-09-02 09:04:36.968806: step 250, lr = 0.0001, loss = 1.60836 (225.6 examples/sec; 0.07091 sec/batch)
2016-09-02 09:04:39.037962: step 260, lr = 0.0001, loss = 1.56989 (212.2 examples/sec; 0.07541 sec/batch)
2016-09-02 09:04:40.491946: step 270, lr = 0.0001, loss = 1.68502 (226.8 examples/sec; 0.07056 sec/batch)
2016-09-02 09:04:42.754090: step 280, lr = 0.0001, loss = 1.65222 (214.8 examples/sec; 0.07450 sec/batch)
2016-09-02 09:04:44.510914: step 290, lr = 0.0001, loss = 1.64218 (241.8 examples/sec; 0.06617 sec/batch)
2016-09-02 09:04:47.301761: step 300, lr = 0.0001, loss = 1.64549 (219.8 examples/sec; 0.07278 sec/batch)
2016-09-02 09:04:50.331641: step 310, lr = 0.0001, loss = 1.56824 (208.1 examples/sec; 0.07689 sec/batch)
2016-09-02 09:04:52.485210: step 320, lr = 0.0001, loss = 1.49187 (226.4 examples/sec; 0.07067 sec/batch)
2016-09-02 09:04:53.878589: step 330, lr = 0.0001, loss = 1.57825 (216.5 examples/sec; 0.07389 sec/batch)
2016-09-02 09:04:55.613968: step 340, lr = 0.0001, loss = 1.57355 (227.0 examples/sec; 0.07047 sec/batch)
2016-09-02 09:04:57.887825: step 350, lr = 0.0001, loss = 1.58644 (231.2 examples/sec; 0.06920 sec/batch)
2016-09-02 09:05:00.791740: step 360, lr = 0.0001, loss = 1.57324 (224.1 examples/sec; 0.07139 sec/batch)
2016-09-02 09:05:03.385293: step 370, lr = 0.0001, loss = 1.56881 (214.0 examples/sec; 0.07476 sec/batch)
2016-09-02 09:05:05.684695: step 380, lr = 0.0001, loss = 1.57652 (223.3 examples/sec; 0.07166 sec/batch)
2016-09-02 09:05:07.671139: step 390, lr = 0.0001, loss = 1.59098 (219.5 examples/sec; 0.07288 sec/batch)
2016-09-02 09:05:09.384184: step 400, lr = 0.0001, loss = 1.57292 (222.9 examples/sec; 0.07177 sec/batch)
2016-09-02 09:05:12.028866: step 410, lr = 0.0001, loss = 1.59291 (224.0 examples/sec; 0.07144 sec/batch)
2016-09-02 09:05:13.912951: step 420, lr = 0.0001, loss = 1.56065 (206.7 examples/sec; 0.07741 sec/batch)
2016-09-02 09:05:15.567746: step 430, lr = 0.0001, loss = 1.48168 (220.1 examples/sec; 0.07270 sec/batch)
2016-09-02 09:05:17.913313: step 440, lr = 0.0001, loss = 1.52358 (220.5 examples/sec; 0.07256 sec/batch)
2016-09-02 09:05:20.215508: step 450, lr = 0.0001, loss = 1.37813 (163.3 examples/sec; 0.09800 sec/batch)
2016-09-02 09:05:21.685596: step 460, lr = 0.0001, loss = 1.66846 (237.3 examples/sec; 0.06741 sec/batch)
2016-09-02 09:05:24.225214: step 470, lr = 0.0001, loss = 1.70520 (219.6 examples/sec; 0.07287 sec/batch)
2016-09-02 09:05:26.203767: step 480, lr = 0.0001, loss = 1.58498 (234.3 examples/sec; 0.06827 sec/batch)
2016-09-02 09:05:28.258299: step 490, lr = 0.0001, loss = 1.50158 (226.4 examples/sec; 0.07067 sec/batch)

two machines:
NFO:tensorflow:Finished running Summary operation.
INFO:tensorflow:Worker 0: 2016-09-01 17:16:20.459047: step 12450, loss = 0.66(21.7 examples/sec; 0.737  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:16:43.294597: step 12480, loss = 0.50(20.7 examples/sec; 0.773  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:17:06.094232: step 12510, loss = 0.54(21.8 examples/sec; 0.733  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:17:29.405047: step 12540, loss = 0.51(20.5 examples/sec; 0.781  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:17:52.529421: step 12570, loss = 0.37(20.8 examples/sec; 0.768  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:18:15.177700: step 12600, loss = 0.49(21.3 examples/sec; 0.750  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:18:38.046958: step 12630, loss = 0.43(21.9 examples/sec; 0.731  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:19:00.978593: step 12660, loss = 0.48(20.7 examples/sec; 0.773  sec/batch)
INFO:tensorflow:Running Summary operation on the chief.
INFO:tensorflow:Finished running Summary operation.
INFO:tensorflow:Worker 0: 2016-09-01 17:19:24.747284: step 12690, loss = 0.47(20.7 examples/sec; 0.772  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:20:10.874324: step 12750, loss = 0.58(20.3 examples/sec; 0.788  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:20:33.881611: step 12780, loss = 0.49(21.0 examples/sec; 0.762  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:20:56.813524: step 12810, loss = 0.63(21.2 examples/sec; 0.755  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:21:19.631475: step 12840, loss = 0.36(21.0 examples/sec; 0.763  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:21:42.239483: step 12870, loss = 0.30(20.7 examples/sec; 0.773  sec/batch)
INFO:tensorflow:Worker 0: 2016-09-01 17:22:04.912129: step 12900, loss = 0.75(21.3 examples/sec; 0.750  sec/batch)

**Comment by poxvoculi** [comment]
You haven't said anything specific about the structure of your model, or exactly what about this situation surprises you.  

Many or most models will run faster on a single GPU than they will when split across multiple GPUs, if it's possible to fit on a single GPU.

Here's why: The latency per step is lower bounded by the most expensive path through a DAG where each node is a kernel execution, whose cost is the time that kernel takes, and each edge is a data dependency, whose cost is the time it takes to transfer the data from the computation that produces it to the computation where it is consumed.  If the model runs completely on one GPU, the data transfer times are all zero, and the overall execution time is just the sum of the kernel execution times.  GPU computations can be very fast, from a few microseconds to a few milliseconds.  By contrast, data transfers by RPC and even via PCIe bus are relatively slow.  

Consider the following example.  We'd like to do a dot product of two large vectors, of 10^9 float32 values, all of which reside on GPU0 to begin with.  We also have GPU1 available, on the same PCIe bus with a bandwidth of 6GB/s.  Naively it seems like partitioning the computation between devices might be a good idea.  Suppose both GPUs have an internal memory bandwidth of 200GB/s.  Based on that limitation, doing the dot product just inside GPU0 should take (10^9 * 4 / 200x10^9) secs = 0.02 sec.  Half of the computation would take 0.01 sec.  But copying 1/2 of the data between GPUs at 6GB/s would take 0.5 * 4 / 6 = 0.33 sec, much longer than the on-GPU computation.  If the data transfer time is even slower, say 10Gbit/sec over a typical RPC network link, then the disadvantage is even greater.

Back to your specific case: It sounds like you're trying a separate ps/worker architecture.  For the reasons just cited, it should be clear that when using a single worker it's going to be slower; this sort of architecture is only advantageous if you use a lot of workers, so that the wall-clock training time is reduced.  

But maybe you're wondering why the time is 10x slower, and yet by your measurement the network is only 40% busy?  There could be lots of reasons.  1Gb/s is a slow network by contemporary standards.  It's usually hard to get more than 80% of the theoretical capacity of a network, and maybe there's other traffic going through it competing with your workload.  There may be other jobs timesharing with your TF jobs on both machines that are competing for kernel resources that interfere with RPC latency.  If you're getting 80% of the theoretical capacity going in each direction, when there's something to send, and measuring overall capacity as bidirectional, then you could see 40% bandwidth consumption when your job is completely network bound.

**Comment by heroleap** [comment]
I am AlvinChen13's colleague and I am going to give more details on the structure. We uses CustomNet CNN to do image classification such as gender detection. We execute our CNN model under the following two cases.

(1) Single Server Case - We mimic the google cifar muti-GPU model to modify our CNN to be able to execute on two GPUs (both GPUs are K40 as mentioned by AlvinChen13 installed on a single server)
The training performance is ~210 examples/sec if we train the CNN on both GPUs and 150 examples/sec if we train CNN only on a single GPU (GPU0 in our case).

(2) Distributed Case: We mimic the google Inception model to modify our CNN to be able to execute on one server with one PS and one worker (worker uses only one GPU for training).
The performance is almost halved (~60 examples/sec).
Then, we further extend our case, in which we use one PS and two workers. PS and one worker is placed on one server and the other worker is placed on another server. Two servers are connected by Gigabyte network switch. During training, we disabled almost all other network traffic within the network. 
The performance is getting worse again, where worker0 and worker1 are both trained at ~20 examples/sec. (We use sync training method)

The number surprised us. From our experiment, the performance keeps getting worse if we add more servers. In our case, single server with multiple GPU trains at hundreds examples/sec. One worker trains at 60 examples/sec (worse than single server case). Two workers train at 40 examples/sec combined (20 examples/sec * 2 workers) which is even worse than single worker. Do you have any idea on how it happens?

**Comment by poxvoculi** [comment]
Most likely, somewhere, some resource is periodically over-committed, causing everything else to wait.  You might try using the tools described here #1824 
to look at the time intervals between operations.  However, you could also use other tools to monitor memory use (is anything about your setup causing swapping), hack the code to time individual RPCs and operations at the PS server, and other techniques to look for operations that violate your assumptions about how they should be behaving.  If you're doing synchronous steps, look for which operation is last to complete at join points, and try to figure out why.

**Comment by JianbangZ** [comment]
@heroleap 
I think I am having the exact same problem as yours, and my approach is also similar to yours. BTW, I am now using 40Gb network, I do notice some performance increase compare to 1G (improve from 10 times slower to maybe 4 times slower), and my network usage is only about 10% I believe. There could be a lot of different reasons to cause this slowness. It's good if we can do some joint experiment and figure this out together?

**Comment by wangyongliang** [comment]
@AlvinChen13 
Hi Alvin~

I have tried distributed inception model in https://github.com/tensorflow/models.git with 1 worker，2 workers，4 workers on GPU(K20C) server without infinityband.
compared with single GPU version, distributed looks like can't reduce any training time to achieve the same error rate.
